Unit test failure: get_subjectgroup_observations

```/Users/yunwu/MEP/ecoscope-workflows/tests/test_tasks.py::test_consumes_and_or_produces_dataframe[get_subjectgroup_observations] failed: tf = TaskFixture(task=DistributedTask(func=<function get_subjectgroup_observations at 0x147c21e10>, arg_prevalidators={}, r...'/Users/yunwu/MEP/ecoscope-workflows/ecoscope_workflows/tasks/io/get-subjectgroup-observations.example-return.parquet')
tmp_path = PosixPath('/private/var/folders/b8/nns96jhx4j7222yjslzrhnnc0000gn/T/pytest-of-yunwu/pytest-6/test_consumes_and_or_produces_0')

    @pytest.mark.parametrize(
        "tf",
        task_fixtures.values(),
        ids=task_fixtures.keys(),
    )
    def test_consumes_and_or_produces_dataframe(
        tf: TaskFixture,
        tmp_path: Path,
    ):
        args = (
            (gpd.read_parquet(tf.example_input_dataframe_path),)
            if tf.example_input_dataframe_path
            else ()  # io tasks don't have input dataframes
        )
>       in_memory = tf.task(*args, **tf.kws)

tests/test_tasks.py:149: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
ecoscope_workflows/decorators.py:123: in __call__
    else func_copy(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

client = 'MEP_DEV', subject_group_name = 'Elephants', include_inactive = True
since = datetime.datetime(2011, 1, 1, 0, 0)
until = datetime.datetime(2023, 1, 1, 0, 0)

    @distributed(tags=["io"])
    def get_subjectgroup_observations(
        client: EarthRangerClient,
        subject_group_name: Annotated[
            str, Field(description="Name of EarthRanger Subject")
        ],
        include_inactive: Annotated[
            bool,
            Field(description="Whether or not to include inactive subjects"),
        ],
        since: Annotated[datetime, Field(description="Start date")],
        until: Annotated[datetime, Field(description="End date")],
    ) -> DataFrame[SubjectGroupObservationsGDFSchema]:
        """Get observations for a subject group from EarthRanger."""
>       return client.get_subjectgroup_observations(
            subject_group_name=subject_group_name,
            include_subject_details=True,
            include_inactive=include_inactive,
            since=since,
            until=until,
        )
E       AttributeError: 'str' object has no attribute 'get_subjectgroup_observations'

ecoscope_workflows/tasks/io/_earthranger.py:42: AttributeError
```

--- a/ecoscope_workflows/connections.py
+++ b/ecoscope_workflows/connections.py
@@ -32,7 +32,7 @@ class _DataConnection(_Settings):
             pyproject_toml_table_header=(
                 "connections",
                 cls.__ecoscope_connection_type__,
-                name.lower(),
+                name,
             ),
         )
         _cls = type(

--- a/tests/tasks/test_io_earthranger.py
+++ b/tests/tasks/test_io_earthranger.py
@@ -1,0 +1,108 @@
+from dataclasses import dataclass
+from datetime import datetime
+from importlib.resources import files
+from pathlib import Path
+from typing import Callable
+
+import os
+import geopandas as gpd
+import pandas as pd
+import pytest
+
+
+from ecoscope_workflows.connections import EarthRangerConnection
+from ecoscope_workflows.decorators import DistributedTask
+from ecoscope_workflows.tasks.io import get_subjectgroup_observations
+from ecoscope_workflows.serde import gpd_from_parquet_uri
+
+pytestmark = pytest.mark.io
+
+
+@dataclass
+class TaskFixture:
+    task: DistributedTask
+    input_dataframe_arg_name: str
+    example_input_dataframe_path: str
+    kws: dict
+    assert_that_return_dataframe: list[Callable[[gpd.GeoDataFrame], bool]]
+    example_return_path: str
+
+
+task_fixtures = {
+    "get_subjectgroup_observations": pytest.param(
+        TaskFixture(
+            task=get_subjectgroup_observations,
+            input_dataframe_arg_name="",
+            example_input_dataframe_path="",
+            kws=dict(
+                client=EarthRangerConnection(
+                    server=os.environ["EARTHRANGER_SERVER"],
+                    username=os.environ["EARTHRANGER_USERNAME"],
+                    password=os.environ["EARTHRANGER_PASSWORD"],
+                    tcp_limit="5",
+                    sub_page_size="4000",
+                ).get_client(),
+                subject_group_name="Elephants",
+                include_inactive=True,
+                since=datetime.strptime("2011-01-01", "%Y-%m-%d"),
+                until=datetime.strptime("2023-01-01", "%Y-%m-%d"),
+            ),
+            assert_that_return_dataframe=[
+                lambda df: all(
+                    [
+                        col in df
+                        for col in ["geometry", "groupby_col", "fixtime", "junk_status"]
+                    ]
+                ),
+            ],
+            example_return_path=str(
+                files("ecoscope_workflows.tasks.io")
+                / "get-subjectgroup-observations.example-return.parquet"
+            ),
+        )
+    )
+}
+
+
+@pytest.mark.parametrize(
+    "tf",
+    task_fixtures.values(),
+    ids=task_fixtures.keys(),
+)
+def test_consumes_and_or_produces_dataframe(
+    tf: TaskFixture,
+    tmp_path: Path,
+):
+    args = (
+        (gpd.read_parquet(tf.example_input_dataframe_path),)
+        if tf.example_input_dataframe_path
+        else ()  # io tasks don't have input dataframes
+    )
+    in_memory = tf.task(*args, **tf.kws)
+
+    for assert_fn in tf.assert_that_return_dataframe:
+        assert assert_fn(in_memory)
+
+    # we've cached this result for reuse by other tests, so check that cache is not stale
+    cached = gpd.read_parquet(tf.example_return_path)
+    pd.testing.assert_frame_equal(in_memory, cached)
+
+    # compare to `distributed` calling style
+    def serialize_result(gdf: gpd.GeoDataFrame) -> str:
+        path: Path = tmp_path / "result.parquet"
+        gdf.to_parquet(path)
+        return path.as_posix()
+
+    result_path = tf.task.replace(
+        arg_prevalidators=(
+            {tf.input_dataframe_arg_name: gpd_from_parquet_uri}
+            if tf.input_dataframe_arg_name
+            else {}  # io tasks don't have input dataframes
+        ),
+        return_postvalidator=serialize_result,
+        validate=True,
+    )(tf.example_input_dataframe_path, **tf.kws)
+    # this time, the result is written to a file, so we need to read it back in
+    distributed_result = gpd.read_parquet(result_path)
+    # and ensure it's the same as the in-memory result
+    pd.testing.assert_frame_equal(in_memory, distributed_result)

--- a/tests/tasks/test_io_earthranger.py
+++ b/tests/tasks/test_io_earthranger.py
@@ -1,108 +1,37 @@
-from dataclasses import dataclass
 from datetime import datetime
 from importlib.resources import files
-from pathlib import Path
-from typing import Callable
 
 import os
 import geopandas as gpd
 import pandas as pd
 import pytest
 
-
 from ecoscope_workflows.connections import EarthRangerConnection
-from ecoscope_workflows.decorators import DistributedTask
 from ecoscope_workflows.tasks.io import get_subjectgroup_observations
-from ecoscope_workflows.serde import gpd_from_parquet_uri
 
 pytestmark = pytest.mark.io
 
 
-@dataclass
-class TaskFixture:
-    task: DistributedTask
-    input_dataframe_arg_name: str
-    example_input_dataframe_path: str
-    kws: dict
-    assert_that_return_dataframe: list[Callable[[gpd.GeoDataFrame], bool]]
-    example_return_path: str
+def test_get_subject_group_observations():
+    result = get_subjectgroup_observations(
+        client=EarthRangerConnection(
+            server=os.environ["EARTHRANGER_SERVER"],
+            username=os.environ["EARTHRANGER_USERNAME"],
+            password=os.environ["EARTHRANGER_PASSWORD"],
+            tcp_limit="5",
+            sub_page_size="4000",
+        ).get_client(),
+        subject_group_name="Elephants",
+        include_inactive=True,
+        since=datetime.strptime("2011-01-01", "%Y-%m-%d"),
+        until=datetime.strptime("2023-01-01", "%Y-%m-%d"),
+    )
 
-
-task_fixtures = {
-    "get_subjectgroup_observations": pytest.param(
-        TaskFixture(
-            task=get_subjectgroup_observations,
-            input_dataframe_arg_name="",
-            example_input_dataframe_path="",
-            kws=dict(
-                client=EarthRangerConnection(
-                    server=os.environ["EARTHRANGER_SERVER"],
-                    username=os.environ["EARTHRANGER_USERNAME"],
-                    password=os.environ["EARTHRANGER_PASSWORD"],
-                    tcp_limit="5",
-                    sub_page_size="4000",
-                ).get_client(),
-                subject_group_name="Elephants",
-                include_inactive=True,
-                since=datetime.strptime("2011-01-01", "%Y-%m-%d"),
-                until=datetime.strptime("2023-01-01", "%Y-%m-%d"),
-            ),
-            assert_that_return_dataframe=[
-                lambda df: all(
-                    [
-                        col in df
-                        for col in ["geometry", "groupby_col", "fixtime", "junk_status"]
-                    ]
-                ),
-            ],
-            example_return_path=str(
-                files("ecoscope_workflows.tasks.io")
-                / "get-subjectgroup-observations.example-return.parquet"
-            ),
+    cached = gpd.read_parquet(
+        str(
+            files("ecoscope_workflows.tasks.io")
+            / "get-subjectgroup-observations.example-return.parquet"
         )
     )
-}
 
-
-@pytest.mark.parametrize(
-    "tf",
-    task_fixtures.values(),
-    ids=task_fixtures.keys(),
-)
-def test_consumes_and_or_produces_dataframe(
-    tf: TaskFixture,
-    tmp_path: Path,
-):
-    args = (
-        (gpd.read_parquet(tf.example_input_dataframe_path),)
-        if tf.example_input_dataframe_path
-        else ()  # io tasks don't have input dataframes
-    )
-    in_memory = tf.task(*args, **tf.kws)
-
-    for assert_fn in tf.assert_that_return_dataframe:
-        assert assert_fn(in_memory)
-
-    # we've cached this result for reuse by other tests, so check that cache is not stale
-    cached = gpd.read_parquet(tf.example_return_path)
-    pd.testing.assert_frame_equal(in_memory, cached)
-
-    # compare to `distributed` calling style
-    def serialize_result(gdf: gpd.GeoDataFrame) -> str:
-        path: Path = tmp_path / "result.parquet"
-        gdf.to_parquet(path)
-        return path.as_posix()
-
-    result_path = tf.task.replace(
-        arg_prevalidators=(
-            {tf.input_dataframe_arg_name: gpd_from_parquet_uri}
-            if tf.input_dataframe_arg_name
-            else {}  # io tasks don't have input dataframes
-        ),
-        return_postvalidator=serialize_result,
-        validate=True,
-    )(tf.example_input_dataframe_path, **tf.kws)
-    # this time, the result is written to a file, so we need to read it back in
-    distributed_result = gpd.read_parquet(result_path)
-    # and ensure it's the same as the in-memory result
-    pd.testing.assert_frame_equal(in_memory, distributed_result)
+    pd.testing.assert_frame_equal(result, cached)

--- a/tests/test_connections.py
+++ b/tests/test_connections.py
@@ -84,16 +84,17 @@ def mock_toml_config(tmp_path: Path):
 
 def test_connection_named_from_toml(mock_toml_config: Path):
     with patch("ecoscope_workflows.config.PATH", mock_toml_config):
-        conn = EarthRangerConnection.from_named_connection("mep_dev")
-        assert conn.server == "https://mep-dev.pamdas.org"
-        assert conn.username == "user"
+        with patch.dict(os.environ, clear=True):
+            conn = EarthRangerConnection.from_named_connection("mep_dev")
+            assert conn.server == "https://mep-dev.pamdas.org"
+            assert conn.username == "user"
 
-        assert isinstance(conn.password, SecretStr)
-        assert str(conn.password) == "**********"
-        assert conn.password.get_secret_value() == "pass"
+            assert isinstance(conn.password, SecretStr)
+            assert str(conn.password) == "**********"
+            assert conn.password.get_secret_value() == "pass"
 
-        assert conn.tcp_limit == 5
-        assert conn.sub_page_size == 4000
+            assert conn.tcp_limit == 5
+            assert conn.sub_page_size == 4000
 
 
 @pytest.fixture
@@ -114,25 +115,28 @@ def mock_toml_config_no_secrets(tmp_path: Path):
 
 
 def test_connection_named_from_toml_with_env_secrets(mock_toml_config_no_secrets: Path):
-    with patch("ecoscope_workflows.config.PATH", mock_toml_config_no_secrets):
-        with pytest.raises(ValidationError, match="password"):
-            _ = EarthRangerConnection.from_named_connection("mep_dev")
+    with patch.dict(os.environ, clear=True):
+        with patch("ecoscope_workflows.config.PATH", mock_toml_config_no_secrets):
+            with pytest.raises(ValidationError, match="password"):
+                _ = EarthRangerConnection.from_named_connection("mep_dev")
 
-    with patch("ecoscope_workflows.config.PATH", mock_toml_config_no_secrets):
-        with patch.dict(
-            os.environ,
-            {"ECOSCOPE_WORKFLOWS__CONNECTIONS__EARTHRANGER__MEP_DEV__PASSWORD": "pass"},
-        ):
-            conn = EarthRangerConnection.from_named_connection("mep_dev")
-            assert conn.server == "https://mep-dev.pamdas.org"
-            assert conn.username == "user"
+        with patch("ecoscope_workflows.config.PATH", mock_toml_config_no_secrets):
+            with patch.dict(
+                os.environ,
+                {
+                    "ECOSCOPE_WORKFLOWS__CONNECTIONS__EARTHRANGER__MEP_DEV__PASSWORD": "pass"
+                },
+            ):
+                conn = EarthRangerConnection.from_named_connection("mep_dev")
+                assert conn.server == "https://mep-dev.pamdas.org"
+                assert conn.username == "user"
 
-            assert isinstance(conn.password, SecretStr)
-            assert str(conn.password) == "**********"
-            assert conn.password.get_secret_value() == "pass"
+                assert isinstance(conn.password, SecretStr)
+                assert str(conn.password) == "**********"
+                assert conn.password.get_secret_value() == "pass"
 
-            assert conn.tcp_limit == 5
-            assert conn.sub_page_size == 4000
+                assert conn.tcp_limit == 5
+                assert conn.sub_page_size == 4000
 
 
 def test_resolve_client_from_env(named_mock_env):

--- a/tests/test_tasks.py
+++ b/tests/test_tasks.py
@@ -1,5 +1,4 @@
 from dataclasses import dataclass
-from datetime import datetime
 from importlib.resources import files
 from pathlib import Path
 from typing import Callable
@@ -8,11 +7,9 @@ import geopandas as gpd
 import pandas as pd
 import pytest
 
-from ecoscope_workflows.connections import EarthRangerConnection
 from ecoscope_workflows.decorators import DistributedTask
 from ecoscope_workflows.serde import gpd_from_parquet_uri
 from ecoscope_workflows.tasks.analysis import calculate_time_density
-from ecoscope_workflows.tasks.io import get_subjectgroup_observations
 from ecoscope_workflows.tasks.preprocessing import (
     process_relocations,
     relocations_to_trajectory,
@@ -30,35 +27,6 @@ class TaskFixture:
 
 
 task_fixtures = {
-    "get_subjectgroup_observations": pytest.param(
-        TaskFixture(
-            task=get_subjectgroup_observations,
-            input_dataframe_arg_name="",
-            example_input_dataframe_path="",
-            kws=dict(
-                client=EarthRangerConnection.from_named_connection(
-                    "MEP_DEV"
-                ).get_client(),
-                subject_group_name="Elephants",
-                include_inactive=True,
-                since=datetime.strptime("2011-01-01", "%Y-%m-%d"),
-                until=datetime.strptime("2023-01-01", "%Y-%m-%d"),
-            ),
-            assert_that_return_dataframe=[
-                lambda df: all(
-                    [
-                        col in df
-                        for col in ["geometry", "groupby_col", "fixtime", "junk_status"]
-                    ]
-                ),
-            ],
-            example_return_path=str(
-                files("ecoscope_workflows.tasks.io")
-                / "get-subjectgroup-observations.example-return.parquet"
-            ),
-        ),
-        marks=pytest.mark.io,
-    ),
     "process_relocations": TaskFixture(
         task=process_relocations,
         input_dataframe_arg_name="observations",

--- a/tests/test_tasks.py
+++ b/tests/test_tasks.py
@@ -8,6 +8,7 @@ import geopandas as gpd
 import pandas as pd
 import pytest
 
+from ecoscope_workflows.connections import EarthRangerConnection
 from ecoscope_workflows.decorators import DistributedTask
 from ecoscope_workflows.serde import gpd_from_parquet_uri
 from ecoscope_workflows.tasks.analysis import calculate_time_density
@@ -35,7 +36,7 @@ task_fixtures = {
             input_dataframe_arg_name="",
             example_input_dataframe_path="",
             kws=dict(
-                client="MEP_DEV",
+                client=EarthRangerConnection.from_named_connection("MEP_DEV").get_client(),
                 subject_group_name="Elephants",
                 include_inactive=True,
                 since=datetime.strptime("2011-01-01", "%Y-%m-%d"),

--- a/tests/test_tasks.py
+++ b/tests/test_tasks.py
@@ -36,7 +36,9 @@ task_fixtures = {
             input_dataframe_arg_name="",
             example_input_dataframe_path="",
             kws=dict(
-                client=EarthRangerConnection.from_named_connection("MEP_DEV").get_client(),
+                client=EarthRangerConnection.from_named_connection(
+                    "MEP_DEV"
+                ).get_client(),
                 subject_group_name="Elephants",
                 include_inactive=True,
                 since=datetime.strptime("2011-01-01", "%Y-%m-%d"),

